src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -894,114 +961,155 @@ namespace stan {
-      // allow int -> double promotion
-      bool types_compatible
-        = lhs_base_type == rhs_base_type
-        || (lhs_base_type == DOUBLE_T && rhs_base_type == INT_T);
-      if (!types_compatible) {
-        error_msgs << "base type mismatch in assignment"
-                   << "; variable name = "
-                   << a.var_dims_.name_
-                   << ", type = ";
-        write_base_expr_type(error_msgs, lhs_base_type);
-        error_msgs << "; right-hand side type=";
-        write_base_expr_type(error_msgs, rhs_base_type);
+      // no compound assign for array types  (std::vector)
+      std::string op_equals = ca.op_;
+      ca.op_ = op_equals.substr(0, op_equals.size()-1);
+      if (inferred_lhs_type.num_dims() > 0) {
+        error_msgs << "Cannot apply operator '" << op_equals
+                   << "' to array variable; variable name = "
+                   << name
+                   << ".";
         error_msgs << std::endl;
         pass = false;
         return;
       }
 
-      if (lhs_type.num_dims_ != a.expr_.expression_type().num_dims_) {
-        error_msgs << "dimension mismatch in assignment"
-                   << "; variable name = "
-                   << a.var_dims_.name_
-                   << ", num dimensions given = "
-                   << lhs_type.num_dims_
-                   << "; right-hand side dimensions = "
-                   << a.expr_.expression_type().num_dims_
+      expr_type lhs_type = inferred_lhs_type.type();
+      expr_type rhs_type = ca.expr_.expression_type();
+      if (lhs_type.is_primitive()
+          && boost::algorithm::starts_with(ca.op_, ".")) {
+        error_msgs << "Cannot apply element-wise operation to scalar"
+                   << "; compound operator is: " << op_equals
                    << std::endl;
         pass = false;
         return;
       }
+      if (lhs_type.is_primitive()
+          && rhs_type.is_primitive()
+          && (lhs_type == DOUBLE_T || lhs_type == rhs_type)) {
+        // done checking <prim> <op>= <prim>
+        pass = true;
+        return;
+      }
 
+      bool types_compatible =
+        // container types allow infix and element-wise operations
+        // when lhs and rhs are same shape, and broadcast operations
+        // when rhs is double and lhs is vector, row_vector, or matrix
+        (lhs_type == rhs_type
+         || (lhs_type == VECTOR_T && rhs_type == DOUBLE_T)
+         || (lhs_type == ROW_VECTOR_T && rhs_type == DOUBLE_T)
+         || (lhs_type == ROW_VECTOR_T && rhs_type == MATRIX_T)
+         || (lhs_type == MATRIX_T && rhs_type == DOUBLE_T));
+      if (!types_compatible) {
+        error_msgs << "Cannot apply operator '" << op_equals
+                   << "' to operands;"
+                   << " left-hand side type = " << lhs_type
+                   << "; right-hand side type=" << rhs_type
+                   << std::endl;
+        pass = false;
+        return;
+      }
+      std::string op_name;
```

Comment:
You can just set `ca.op_name_` directly here and save the big gap between when `op_name` is declared, where it's defined, and where it's used.